README.md
# turkish-suffixes
_**A PHP/C# Class & JS Functions for proper noun and number suffixes in Turkish Language.**_

Should work with PHP >=5, .NET >=2.0 and with all browsers

_Not tested with PHP 7!_

_(May not work as perfect as a Turkish/Literature teacher, because the Turkish Language is very complicated when using suffixes.)_

--

_**Türkçe'deki özel isim ve sayılardaki hal ekleri (ve dahi bağlacı) için bir PHP Sınıfı.**_

PHP >=5'te, .NET >=2.0'da ve tüm tarayıcılarda çalışıyor olmalı.

_PHP 7 ile test edilmedi!_

_(Lütfen tüm isimlerde kusursuz sonuç veremeyeceğini unutmayın. Türkçe zor bir dil ve aslında bu ekler getirilirken ismin tamlama olup olmamasına da bakılması gerekiyor. Yapmış olduğumuz sınıf, ismin tamlama olup olmadığını sadece tahmin etmeye çalışıyor; ancak, haliyle, tüm isimlerle test etmedim.)_

# Usage | Kullanım

_For full usage example and performance info, please see the demo files or check http://siis.com.tr/turkish-suffixes page._

It's actually very easy, just include Turkce to you project, like:

_PHP_
`require_once("Turkce.php");`

_JavaScript_
`<script src="Turkce.min.js"></script>`

and then use it for a name like this (all functions are static for PHP/C#):

_PHP_
`Turkce::bulunmaHali("Utku");` or `Turkce::locativeCase("Utku");`

_C#_
`Turkce.bulunmaHali("Utku");` or `Turkce.locativeCase("Utku");`

_JavaScript_
`Turkce.bulunmaHali("Utku");` or `Turkce.locativeCase("Utku");`

and Turkce Class/Functions returns "Utku" as:

`Utku'da`

That's all folks!

--

_Lütfen tam kullanım örnekleri ve performans bilgisi için demo dosyalarına ya da http://siis.com.tr/turkish-suffixes sayfasına göz atın._

Aslında oldukça basit, sadece Turkce'yi projenize dahil edin, mesela:

_PHP_
`require_once("Turkce.php");`

_JavaScript_
`<script src="Turkce.js"></script>`

sonra bir isme uygulamak için, şu şekilde kullanın (PHP/C#'ta tüm fonksiyonlar statik):

_PHP_
`Turkce::bulunmaHali("Utku");` veya `Turkce::locativeCase("Utku");`

_C#_
`Turkce.bulunmaHali("Utku");` veya `Turkce.locativeCase("Utku");`

_JavaScript_
`Turkce.bulunmaHali("Utku");` veya `Turkce.locativeCase("Utku");`

ve Turkce sınıfı "Utku" ismini şu hale getirsin:

`Utku'da`

Bu kadar basit!

# For What? | Neden?

For: `Ali'yi silmek istediğinizden emin misiniz?` or `Ali'nin karnı acıktı.`

Not: `Ali adlı kullanıcıyı silmek istediğinizden emin misiniz?` or `Ali adlı kullanıcının karnı acıktı.`

Or not: `Şu kullanıcıyı silmek istediğinize emin misiniz: Ali` or `Karnı acıktı: Ali`

Briefly: For proper usage of Turkish.

--

Bunun için: `Ali'yi silmek istediğinizden emin misiniz?` veya `Ali'nin karnı acıktı.`

Şu olmasın diye: `Ali adlı kullanıcıyı silmek istediğinizden emin misiniz?` veya  `Ali adlı kullanıcının karnı acıktı.`

Ya da şu: `Şu kullanıcıyı silmek istediğinize emin misiniz: Ali` veya  `Karnı acıktı: Ali`

Özetle: Türkçe'nin daha doğru kullanımı için.

# Bugs | Sorunlar

Didn't noticed any wrong suffix so far; but if you find, please open an issue at least.

--

Şu ana kadar yanlış ek geldiğini görmedik; fakat bulursanız, lütfen en azından bir "issue" oluşturun.


# Notes | Notlar

Not well tested C# Class. It may have performance issues. Please don't hesitate to report bugs.

--

C# sınıfı çok iyi test edilmedi. Performans sorunları olabilir. Lütfen sorunları bildirmekten çekinmeyiniz.

# Thanks | Teşekkür

Thank you, for your interest. Thank you future contributors, for your future contributions. Thank you Müge Bora Bayraktar, for helping me.

--

Öncelikle ilgin için sana teşekkür ederiz, buraya kadar yorduk seni.
Ardından gelecekte katkı sağlayacak kişilere, sağlayacakları katkılar için teşekkür ederiz.
Son olarak da Müge Bora Bayraktar'a yardımları için teşekkür ederiz.


CHANGELOG.md
## 1.4.1 (2016-10-07)
Features:

- Added full-js-demo.html (http://siis.com.tr/turkish-suffixes page) for detailed usage and demo.

Bugfixes:

- Fixed ordinalNumber function in JS.

## 1.4 (2016-10-06)
Features:

- Added pronounce check (with hashtag and mention support) for some brands, people, products, etc. for proper suffixes.

Bugfixes:

- Fixed suffix letter for genitive case in JS.
- Fixed problems with distributive and ordinal number methods in C#.

## 1.3 (2016-10-05)

Features:

- Added distributive and ordinal number methods.
- added support for JavaScript🤘🏻
- added support for C#🤘🏻

Bugfixes:

- Fixed a problem for some short words which are contains Turkish letters.
- Added support for "-ığı" suffix.

Changes:

- some long and repetitive regex patterns defined as variables.
- changed $sh variable name to $vowels
- updated readme file.

## 1.2 (2016-09-30)

Features:

- added support for numbers.
- added support for pronounce usage. You can use it like this (last 3 letters of pronounce would be enough): `Turkce::bulunmaHali("epıl", "Apple");` or `Turkce::locativeCase("epıl", "Apple");`
- added .gitignore file.

Bugfixes:

- Tested for PHP's UTF-8 case-insensitive bug and it's working well

Performance:

- Removed logical control for predefined suffix letters.

Changes:

- updated readme file.
- updated demo file for pronounce usage example.

## 1.1 (2016-01-02)

Features:

- added support to use methods with English method names (like `Turkce::locativeCase("Noun");`).

Bugfixes:

- fixed an error for nouns with last 3 letters are consonant (and this gives a little extra performance =)).

Changes:

- updated variable names to English.
- updated readme file.

## 1.0 (2016-02-14)

- initial commit.


js/Turkce.js
/**
 * Created by IntelliJ IDEA.
 * Author: A. Utku Sipahioğlu (http://teorikdeli.com)
 * Company: Siis (http://siis.com.tr | http://siisgames.com)
 *
 * License: MIT
 * https://github.com/SiisOfficial/turkish-suffixes
 *
 * Demo & detailed usage: http://siis.com.tr/turkish-suffixes
 *
 * Version: 1.4.2
 */
var Turkce = {
    /*
     |--------------------------------------------------------------------------
     | Turkish Case Suffixes For JavaScript
     |--------------------------------------------------------------------------
     |
     |   Should work with all browsers.
     |
     |   --------
     |
     |   Tüm tarayıcılarda çalışıyor olmalı.
     |
     */

    nounPhraseEnds: /oğlu$|esi$|aşı$|işi$|eşi$|ağı$|isi$|iği$|ığı$|ıgı$|oglu$|okulu$|odası$|leri$|ları$|lari$/i,
    vowel: /[oueiöüaıOUEİÖÜAI]/,
    vowelE: /[oueiöüaıOUEİÖÜAI]$/,
    vowelG: /[oueiöüaıOUEİÖÜAI]/g,

    //  This is the simplest & fastest solution for now.
    //  I really don't want to check last noun of given var noun,
    //  bacause (for example vive) last noun can be a Turkish brand or something.
    //  I'd suggest you to use var fake (like Turkce.accusativeCase("yıv", "view");).
    //
    //  Only for some brands, people, products, etc.
    pronounces: {
        'apple': 'pıl',
        'google': 'gıl',
        'twitter': 'tır',
        'youtube': 'tup',
        'chrome': 'rom',
        'developer': 'pır',
        'appledeveloper': 'pır',
        'googledeveloper': 'pır',
        'facebookdeveloper': 'pır',
        'googlecardboard': 'ord',
        'cardboard': 'ord',
        'iphone': 'fon',
        'ipad': 'ped',
        'watch': 'voç',
        'appleiphone': 'fon',
        'appleipad': 'ped',
        'applewatch': 'voç',
        'imessage': 'sıc',
        'php': 'çpi',
        'steam': 'tim',
        'valve': 'alv',
        'facebookmessenger': 'cır',
        'messenger': 'cır',
        'snapchat': 'çet',
        'chat': 'çet',
        'skype': 'ayp',
        'unity': 'iti',
        'spotify': 'fay',
        'htc': 'ysi',
        'half-life': 'ayf',
        'halflife': 'ayf',
        'htcvive': 'avy',
        '9gag': 'geg',
        'tv': 'ivi',
        'appletv': 'ivi',
        'usa': 'sey',
        'googlemaps': 'eps',
        'maps': 'eps',
        'cern': 'örn',
        'adobe': 'dob',
        'imdb': 'ibi',
        'italy': 'ali',
        'store': 'tor',
        'appstore': 'tor',
        'playstore': 'tor',
        'elonmusk': 'ask',
        'einstein': 'ayn',
        'xcode': 'kod',
        'github': 'hab',
        'bitbucket': 'kıt',
        'slipknot': 'nat',
        'ironmaiden': 'dın',
        'slayer': 'yır',
        'gameofthrones': 'ons',
        'thesimpsons': 'ıns',
        'familyguy': 'gay',
        'starwars': 'ors',
        'darthvader': 'dır',
        'beethoven': 'vın',
        'vr': 'yar',
        'playstationvr': 'yar',
        'gearvr': 'yar',
        'googlevr': 'yar',
        'steamvr': 'yar',
        'osvr': 'yar',
        'vrfirst': 'öst',
        'unrealengine': 'cin',
        'cryengine': 'cin',
        '$': 'lar',
        '€': 'uro',
        '₺': 'ele'
    },

    /**
     * Checks the fake variable if it's undefined.
     * If so, sets it's default value to false.
     *
     * @param v
     * @returns {boolean || string}
     */
    isFake: function(v) {
        return (typeof v == "undefined" ? false : v);
    },

    /**
     * Checks the noun if it has a pronounce in pronounces
     *
     * @param n
     * @returns {boolean || string}
     */
    checkHasPronounce: function(n) {
        n = n.replace(/ /g, "").toLowerCase();
        var pronounce;

        if((n.substr(0, 1) == "#" || n.substr(0, 1) == "@") && (pronounce = Turkce.pronounces[n.substr(1)])) {
            return pronounce;
        } else if(pronounce = Turkce.pronounces[n]) {
            return pronounce;
        }

        return false;
    },

    /**
     * Accusative case
     *
     * @param noun
     * @param fake (integer/string if it's a fake noun): this is for pronounce usage
     * @returns {string}
     */
    belirtmeHali: function(noun, fake) {
        noun = noun.toString().trim();
        fake = Turkce.isFake(fake);
        var fake_;

        if(!fake && (fake_ = Turkce.checkHasPronounce(noun))) {
            fake = noun;
            noun = fake_;
        }

        //  Check the ending if it's a number
        if(/\d$/.test(noun)) return this.processForNumber(this.belirtmeHali, noun);

        var last_letter      = noun.substr(-1, 1),
            last_3_letter    = noun.substr(-3, 3),
            nouns            = noun.split(' '),
            noun_length      = nouns.length,
            last_noun_length = nouns[noun_length - 1].length;

        if(fake !== false) noun = fake;

        var blending = "";
        if((Turkce.vowel.test(last_letter) || (/[çÇ]/.test(last_3_letter) && Turkce.vowelE.test(last_letter))) && !/[ğçĞÇ]/.test(last_letter) && !/^çü|^ÇÜ|^ÇÖ|^çö/.test(last_3_letter)) blending = "y";
        if((Turkce.nounPhraseEnds.test(nouns[noun_length - 1]) && last_noun_length > 6)) blending = "n";

        var vovels = last_3_letter.match(Turkce.vowelG),
            suffix = "i";

        if(vovels == null) return (noun + "'" + blending + suffix);

        if(/[öüÖÜ]/.test(vovels[(vovels.length) - 1])) suffix = "ü";
        if(/[aıAI]/.test(vovels[(vovels.length) - 1])) suffix = "ı";
        if(/[ouOU]/.test(vovels[(vovels.length) - 1])) suffix = "u";

        return (noun + "'" + blending + suffix);
    },
    accusativeCase: function(noun, fake) {
        return this.belirtmeHali(noun, fake);
    },

    /**
     * Dative case
     *
     * @param noun
     * @param fake (integer/string if it's a fake noun): this is for pronounce usage
     *
     * @returns {string}
     */
    yonelmeHali: function(noun, fake) {
        noun = noun.toString().trim();
        fake = Turkce.isFake(fake);
        var fake_;

        if(!fake && (fake_ = Turkce.checkHasPronounce(noun))) {
            fake = noun;
            noun = fake_;
        }

        //  Check the ending if it's a number
        if(/\d$/.test(noun)) return this.processForNumber(this.yonelmeHali, noun);

        var last_letter      = noun.substr(-1, 1),
            last_3_letter    = noun.substr(-3, 3),
            nouns            = noun.split(' '),
            noun_length      = nouns.length,
            last_noun_length = nouns[noun_length - 1].length;

        if(fake !== false) noun = fake;

        var blending = "";
        if((Turkce.vowel.test(last_letter) || (/[çÇ]/.test(last_3_letter) && Turkce.vowelE.test(last_letter))) && !/[ğçĞÇ]/.test(last_letter) && !/^çü|^ÇÜ|^ÇÖ|^çö/.test(last_3_letter)) blending = "y";
        if((Turkce.nounPhraseEnds.test(nouns[noun_length - 1]) && last_noun_length > 6)) blending = "n";

        var vovels = last_3_letter.match(Turkce.vowelG),
            suffix = "e";

        if(vovels == null) return (noun + "'" + blending + suffix);

        if(/[ouaıOUAI]/.test(vovels[(vovels.length) - 1])) suffix = "a";

        return (noun + "'" + blending + suffix);

    },
    dativeCase: function(noun, fake) {
        return this.yonelmeHali(noun, fake);
    },

    /**
     * Locative case
     *
     * @param noun
     * @param fake (integer/string if it's a fake noun): this is for pronounce usage
     *
     * @returns {string}
     */
    bulunmaHali: function(noun, fake) {
        noun = noun.toString().trim();
        fake = Turkce.isFake(fake);
        var fake_;

        if(!fake && (fake_ = Turkce.checkHasPronounce(noun))) {
            fake = noun;
            noun = fake_;
        }

        //  Check the ending if it's a number
        if(/\d$/.test(noun)) return this.processForNumber(this.bulunmaHali, noun);

        var last_letter      = noun.substr(-1, 1),
            last_3_letter    = noun.substr(-3, 3),
            last_2_letter    = noun.substr(-2, 2),
            nouns            = noun.split(' '),
            noun_length      = nouns.length,
            last_noun_length = nouns[noun_length - 1].length;

        if(fake !== false) noun = fake;

        var blending = "";
        if((Turkce.nounPhraseEnds.test(nouns[noun_length - 1]) && last_noun_length > 6)) blending = "n";

        var suffix  = "d",
            suffix_ = "e";
        if(/[pçktfşhPÇKTFŞH]/.test(last_2_letter)) suffix = "t";
        if(/[ğĞ]/.test(last_letter)) suffix = "d";
        if(/ş|s|ç/i.test(last_2_letter)) suffix = "t";
        if(Turkce.vowel.test(last_letter)) suffix = "d";
        if(/şk|ŞK/.test(last_3_letter)) suffix_ = "a";

        var vovels = last_3_letter.match(Turkce.vowelG);

        if(vovels == null) return (noun + "'" + blending + suffix + suffix_);

        if(/[ouaıOUAI]/.test(vovels[(vovels.length) - 1])) suffix_ = "a";
        if(/[öüeiÖÜEİ]/.test(vovels[(vovels.length) - 1])) suffix_ = "e";

        return (noun + "'" + blending + suffix + suffix_);
    },
    locativeCase: function(noun, fake) {
        return this.bulunmaHali(noun, fake);
    },

    /**
     * Ablative case
     *
     * @param noun
     * @param fake (integer/string if it's a fake noun): this is for pronounce usage
     *
     * @returns {string}
     */
    ayrilmaHali: function(noun, fake) {
        noun = noun.toString().trim();
        fake = Turkce.isFake(fake);
        var fake_;

        if(!fake && (fake_ = Turkce.checkHasPronounce(noun))) {
            fake = noun;
            noun = fake_;
        }

        //  Check the ending if it's a number
        if(/\d$/.test(noun)) return this.processForNumber(this.ayrilmaHali, noun);

        var last_letter      = noun.substr(-1, 1),
            last_3_letter    = noun.substr(-3, 3),
            last_2_letter    = noun.substr(-2, 2),
            nouns            = noun.split(' '),
            noun_length      = nouns.length,
            last_noun_length = nouns[noun_length - 1].length;

        if(fake !== false) noun = fake;

        var blending = "";
        if((Turkce.nounPhraseEnds.test(nouns[noun_length - 1]) && last_noun_length > 6)) blending = "n";

        var suffix  = "d",
            suffix_ = "e";
        if(/[pçktfşhPÇKTFŞH]/.test(last_2_letter)) suffix = "t";
        if(/[ğĞ]/.test(last_letter)) suffix = "d";
        if(/ş|s|ç/i.test(last_2_letter)) suffix = "t";
        if(Turkce.vowel.test(last_letter)) suffix = "d";
        if(/şk|ŞK/.test(last_3_letter)) suffix_ = "a";

        var vovels = last_3_letter.match(Turkce.vowelG);

        if(vovels == null) return (noun + "'" + blending + suffix + suffix_ + "n");

        if(/[ouaıOUAI]/.test(vovels[(vovels.length) - 1])) suffix_ = "a";
        if(/[öüeiÖÜEİ]/.test(vovels[(vovels.length) - 1])) suffix_ = "e";

        return (noun + "'" + blending + suffix + suffix_ + "n");
    },
    ablativeCase: function(noun, fake) {
        return this.ayrilmaHali(noun, fake);
    },

    /**
     * Genitive case
     *
     * @param noun
     * @param fake (integer/string if it's a fake noun): this is for pronounce usage
     *
     * @returns {string}
     */
    sahiplikHali: function(noun, fake) {
        noun = noun.toString().trim();
        fake = Turkce.isFake(fake);
        var fake_;

        if(!fake && (fake_ = Turkce.checkHasPronounce(noun))) {
            fake = noun;
            noun = fake_;
        }

        //  Check the ending if it's a number
        if(/\d$/.test(noun)) return this.processForNumber(this.sahiplikHali, noun);

        var last_letter   = noun.substr(-1, 1),
            last_3_letter = noun.substr(-3, 3);

        if(fake !== false) noun = fake;

        var blending = "";
        if(Turkce.vowel.test(last_letter)) blending = "n";

        var vovels = last_3_letter.match(Turkce.vowelG),
            suffix = "i";

        if(vovels == null) return (noun + "'" + blending + suffix + "n");

        if(/[öüÖÜ]/.test(vovels[(vovels.length) - 1])) suffix = "ü";
        if(/[aıAI]/.test(vovels[(vovels.length) - 1])) suffix = "ı";
        if(/[ouOU]/.test(vovels[(vovels.length) - 1])) suffix = "u";

        return (noun + "'" + blending + suffix + "n");
    },
    genitiveCase: function(noun, fake) {
        return this.sahiplikHali(noun, fake);
    },


    /**
     * Comitative case
     *
     * @param noun
     * @param fake (integer/string if it's a fake noun): this is for pronounce usage
     *
     * @returns {string}
     */
    vasitaHali: function(noun, fake) {
        noun = noun.toString().trim();
        fake = Turkce.isFake(fake);
        var fake_;

        if(!fake && (fake_ = Turkce.checkHasPronounce(noun))) {
            fake = noun;
            noun = fake_;
        }

        //  Check the ending if it's a number
        if(/\d$/.test(noun)) return this.processForNumber(this.vasitaHali, noun);

        var last_letter   = noun.substr(-1, 1),
            last_3_letter = noun.substr(-3, 3);

        if(fake !== false) noun = fake;

        var blending = "";
        if(Turkce.vowel.test(last_letter)) blending = "y";

        var vovels = last_3_letter.match(Turkce.vowelG),
            suffix = "e";

        if(vovels == null) return (noun + "'" + blending + "l" + suffix);

        if(/[ouaıOUAI]/.test(vovels[(vovels.length) - 1])) suffix = "a";

        return (noun + "'" + blending + "l" + suffix);
    },
    comitativeCase: function(noun, fake) {
        return this.vasitaHali(noun, fake);
    },

    /**
     * Conjunction  (I'm not sure if this is the correct meaning of this)
     *
     * @param noun
     * @param fake (integer/string if it's a fake noun): this is for pronounce usage
     *
     * @returns {string}
     */
    dahiBaglac: function(noun, fake) {
        noun = noun.toString().trim();
        fake = Turkce.isFake(fake);
        var fake_;

        if(!fake && (fake_ = Turkce.checkHasPronounce(noun))) {
            fake = noun;
            noun = fake_;
        }

        //  Check the ending if it's a number
        if(/\d$/.test(noun)) return this.processForNumber(this.dahiBaglac, noun);

        var last_3_letter = noun.substr(-3, 3);

        if(fake !== false) noun = fake;

        var blending = "d",
            vovels   = last_3_letter.match(Turkce.vowelG),
            suffix   = "e";

        if(vovels == null) return (noun + " " + blending + suffix);

        if(/[ouaıOUAI]/.test(vovels[(vovels.length) - 1])) suffix = "a";

        return (noun + " " + blending + suffix);
    },
    conjunction: function(noun, fake) {
        return this.dahiBaglac(noun, fake);
    },

    /**
     * Ordinal Number
     *
     * @param number
     * @param integer (if it's a integer, use "'", otherwise don't.)
     *
     * !Warning: Don't give integer when you directly call this method, because it'll override the number with your integer.
     *           This is mostly for internal use (see processForNumber())
     *
     * @returns {string}
     */
    siraSayi: function(number, integer) {
        number  = number.toString().trim();
        integer = Turkce.isFake(integer);

        //  Check the ending if it's a number
        if(/\d$/.test(number)) return this.processForNumber(this.siraSayi, number);

        var last_letter   = number.substr(-1, 1);
        var last_3_letter = number.substr(-3, 3);

        //  Check if the ending with "ört", if so make it "örd" (only for "dört")
        if(last_3_letter.toLowerCase() == "ört") number = number.substr(0, number.length - 1) + (last_letter == "t" ? "d" : "D");

        if(integer !== false) number = integer;

        var blending = "i",
            vovels   = last_3_letter.match(Turkce.vowelG),
            suffix   = "i";

        if(vovels == null) return (number + (integer ? "'" : "") + blending + "nc" + suffix);

        if(/[öüÖÜ]/.test(vovels[(vovels.length) - 1])) suffix = "ü";
        if(/[aıAI]/.test(vovels[(vovels.length) - 1])) suffix = "ı";
        if(/[ouOU]/.test(vovels[(vovels.length) - 1])) suffix = "u";

        if(!Turkce.vowel.test(last_letter)) blending = suffix; else blending = "";

        return (number + (integer ? "'" : "") + blending + "nc" + suffix);
    },
    ordinalNumber: function(number, integer) {
        return this.siraSayi(number);
    },

    /**
     * Distributive
     *
     * @param number
     * @param integer (if it's a integer, use "'", otherwise don't.)
     *
     * !Warning: Don't give integer when you directly call this method, because it'll override the number with your integer.
     *           This is mostly for internal use (see processForNumber())
     *
     * @returns {string}
     */
    ulestirme: function(number, integer) {
        number  = number.toString().trim();
        integer = Turkce.isFake(integer);

        //  Check the ending if it's a number
        if(/\d$/.test(number)) return this.processForNumber(this.ulestirme, number);

        var last_letter   = number.substr(-1, 1);
        var last_3_letter = number.substr(-3, 3);

        //  Check if the ending with "ört", if so make it "örd" (only for "dört")
        if(last_3_letter.toLowerCase() == "ört") number = number.substr(0, number.length - 1) + (last_letter == "t" ? "d" : "D");

        if(integer !== false) number = integer;

        var blending = "",
            vovels   = last_3_letter.match(Turkce.vowelG),
            suffix   = "e";

        if(vovels == null) return (number + (integer ? "'" : "") + blending + suffix + "r");

        if(/[ouaıOUAI]/.test(vovels[(vovels.length) - 1])) suffix = "a";

        if(Turkce.vowel.test(last_letter)) blending = "ş";

        return (number + (integer ? "'" : "") + blending + suffix + "r");
    },
    distributive: function(number, integer) {
        return this.ulestirme(number);
    },

    /**
     * This is for the numbers. It returns callback function with given number's pronounce
     *
     * @param callback
     * @param noun
     * @returns {string}
     */
    processForNumber: function(callback, noun) {
        if(typeof callback == "undefined") callback = this.belirtmeHali;

        var last_number = noun.split('/\D/');
        last_number     = last_number[last_number.length - 1];

        if(last_number == 0) {
            // it is directly zero
            return callback("fır", noun);
        }

        //  We don't need a full word, last 3 letters'd be enough
        var endings      = {
            '1': 'bir',
            '2': 'iki',
            '3': 'üç',
            '4': 'ört',
            '5': 'beş',
            '6': 'ltı',
            '7': 'edi',
            '8': 'kiz',
            '9': 'kuz',
            '10': 'on',
            '20': 'rmi',
            '30': 'tuz',
            '40': 'ırk',
            '50': 'lli',
            '60': 'mış',
            '70': 'miş',
            '80': 'sen',
            '90': 'san'
        };
        //  last number's length
        var digit_length = last_number.length;

        //  Check last number
        var last_1_letter = noun.substr(-1, 1),
            last_1_number = last_1_letter.split('/^\D/');
        last_1_number     = last_1_number[last_1_number.length - 1];
        if(last_1_number != "0") return callback(endings[last_1_number], noun);

        //  Check last two numbers
        var last_2_letter = noun.substr(-2, 2),
            last_2_number = last_2_letter.split('/^\D/');
        last_2_number     = last_2_number[last_2_number.length - 1];
        if(last_2_number != "00") return callback(endings[last_2_number], noun);

        //  Check last three numbers
        var last_3_letter = noun.substr(-3, 3),
            last_3_number = last_3_letter.split('/^\D/');
        last_3_number     = last_3_number[last_3_number.length - 1];
        if(last_3_number != "000") return callback("yüz", noun);

        //  Check last four numbers
        var last_4_letter = noun.substr(-4, 4),
            last_4_number = last_4_letter.split('/^\D/');
        last_4_number     = last_4_number[last_4_number.length - 1];
        if(last_4_number != "0000" || digit_length < 7) return callback("bin", noun);

        //  Check for the rest
        if(digit_length >= 7) {
            // if this is 8, 11, 14, etc.
            if((digit_length + 1) % 3 == 0) {
                if(digit_length % 2 == 0) {
                    return callback("yon", noun);
                } else {
                    return callback("yar", noun);
                }
            } else if(digit_length % 3 == 0 || (digit_length + 2) % 3 == 0) {
                if(digit_length % 2 == 0) {
                    return callback("yar", noun);
                } else {
                    return callback("yon", noun);
                }
            }
        }

        return callback(noun);
    }
};

demo/turkish-suffixes-demo.php
<!DOCTYPE html>
<html lang="tr">
<head>
    <meta charset="utf-8">
    <title>Turkish Suffixes PHP Demo</title>
</head>
<body>
<form method="get" action="">
    <input type="text" name="i" value="<?= $_GET['i'] ?>" placeholder="İsim yaz...">
    <button type="submit">wubba lubba dub dub</button>
    <a href="?stress=test">Stress Test</a>
</form>
<?php
require_once("../php/Turkce.php");

$Nouns = array(
    "Siis",
    "Utku Sipahioğlu",
    "Alper Özgün Yeşil",
    "Burak Köksal",
    "Çağlar Şahin",
    "Merve Kurt",
    "Gizem Bor",
    "Anıl Şen",
    "Fatoş Yılmaz",
    "Görkem Karagöz",
    "Ömer Aybakır",
    "Nevcan Yanık",
    "Pelin Sürgit",
    "Fulya Ulutürk",
    "Tuğçe Altuntaş",
    "Tuğçe Şahpaz",
    "Kubilay Sağlam",
    "Teorik Deli",
    "Melis Karabulut",
    "Ümra Yıldırım",
    "Furkan Demirci",
    "Burcu Zülal Türker",
    "Mavi Deniz Yapar",
    "Gizem Pakyürek",
    "Ellis Bexter",
    "Mert Duygun",
    "Hasan Maratuk",
    "Çiğdem Elif Çelik",
    "Yağmur Işık",
    "Soner Altın",
    "Erkan Kerti",
    "Zafer Elçik",
    "Aykut Guncan",
    "Maksude Ozturk",
    "Maksude Oztürk",
    "Nur Kübra özkan",
    "Dilan Kaygusuz",
    "Burak Karabey",
    "Barış Gürbüz",
    "Ulaş Erdegör",
    "Damla Özsargın",
    "Müge Bora Bayraktar",
    "Elif Akgün",
    "Yunus Ocak",
    "Tuğba Boz",
    "Andaç Yaşar",
    "Fatih Uluçay",
    "MercAn Cebe",
    "Cemre Dağ",
    "Duygu Akyürek",
    "Zeynep Ünal",
    "Betül İlkbaş",
    "Buket Başol",
    "Simge Çakır",
    "Merih Akgünay",
    "Gizem Nida Çınar",
    "Birsel Çatık",
    "Can Aslan",
    "Ece Köken",
    "Başak",
    "Doğukan Tüzün",
    "Yağmur",
    "Erkan Pınarbaşı",
    "Betül",
    "Çağlar",
    "Beste Gökoğlu",
    "Işıl",
    "Can",
    "Nuriye",
    "Özlem Ergün",
    "Ayşegül",
    "Flört",
    "Sis",
    "Sena Akti",
    "Sencer Türkiş",
    "Ateş",
    "Anıl Özkeleş",
    "Fatoş",
    "Sina Tanış",
    "Fersun Çelikiş",
    "Neşe",
    "Kadir Aşk",
    "İmran Başbuğ",
    "Başak Savaşçı",
    "Bulut Barutçu",
    "Veli Baltacı",
    "Ege Atıcı",
    "Burcu Öztuna",
    "Belgin Koca",
    "Nesrin Tuna",
    "Ahmet Bolu",
    "Gülçin Sarı",
    "Zehra Gül Konya",
    "İsmail Bora",
    "Gönenç Ünlü",
    "Elif İrem Koç",
    "Esra Aydoğdu",
    "Mustafa Özaytaç",
    "Ümra",
    "Büşra",
    "Buğra",
    "Su",
    "Cüneyt Gelibolu",
    "İzmir Üniversitesi",
    "İzmir",
    "Üniversite",
    "Mustafa Kemal Atatürk",
    "Türkiye",
    "Amerika Birleşik Devletleri",
    "Birleşik Krallık",
    "World Wide Web",
    "Apple",
    "World",
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    20,
    30,
    40,
    50,
    60,
    70,
    80,
    90,
    100,
    1000,
    1000000,
    1000000000,
    1000000000000
);

if(isset($_GET['i'])) {
    $noun = ucwords($_GET['i']);    //  !Warning: There is a bug for "i" letter when uppercase it to "İ". PHP makes "i" to "I".

    $first_time = preg_split("/ /", microtime());
    $first_time = $first_time[0] + $first_time[1];

    makeBlock($noun);

    $last_time = preg_split("/ /", microtime());
    $last_time = $last_time[0] + $last_time[1];

    echo "<p>" . ($last_time - $first_time) . " seconds for " . $noun . "</p>";
}

if(isset($_GET['stress'])) {
    $first_time = preg_split("/ /", microtime());
    $first_time = $first_time[0] + $first_time[1];

    for($i = 0; $i < count($Nouns); $i++) {
        $noun = ucwords($Nouns[$i]);   //  !Warning: There is a bug for "i" letter when uppercase it to "İ". PHP makes "i" to "I".
        makeBlock($noun);
    }

    $last_time = preg_split("/ /", microtime());
    $last_time = $last_time[0] + $last_time[1];

    echo "<p>" . ($last_time - $first_time) . " seconds for <b>" . count($Nouns) . "</b> names</p>";
}

function makeBlock($noun) {
    if($noun == "World") {
        echo "<blockquote>";
        echo "<b>Belirtme (Accusative) :</b> " . Turkce::accusativeCase("öld", $noun) . "<br>";
        echo "<b>Yönelme (Dative) :</b> " . Turkce::yonelmeHali("öld", $noun) . "<br>";
        echo "<b>Bulunma (Locative) :</b> " . Turkce::bulunmaHali("öld", $noun) . "<br>";
        echo "<b>Ayrılma (Ablative) :</b> " . Turkce::ayrilmaHali("öld", $noun) . "<br>";
        echo "<b>Sahiplik (Genitive) :</b> " . Turkce::sahiplikHali("öld", $noun) . "<br>";
        echo "<b>Vasıta (Comitative) :</b> " . Turkce::vasitaHali("öld", $noun) . "<br>";
        echo "<b>Dahi Bağlacı (Conjunction) :</b> " . Turkce::dahiBaglac("öld", $noun) . "<br>";
        echo "<b>Sıra Sayı (Ordinal Number) :</b> " . Turkce::siraSayi($noun) . "<br>";
        echo "<b>Üleştirme (Distributive) :</b> " . Turkce::ulestirme($noun) . "<br>";
        echo "</blockquote>";
    } else {
        echo "<blockquote>";
        echo "<b>Belirtme (Accusative) :</b> " . Turkce::accusativeCase($noun) . "<br>";
        echo "<b>Yönelme (Dative) :</b> " . Turkce::yonelmeHali($noun) . "<br>";
        echo "<b>Bulunma (Locative) :</b> " . Turkce::bulunmaHali($noun) . "<br>";
        echo "<b>Ayrılma (Ablative) :</b> " . Turkce::ayrilmaHali($noun) . "<br>";
        echo "<b>Sahiplik (Genitive) :</b> " . Turkce::sahiplikHali($noun) . "<br>";
        echo "<b>Vasıta (Comitative) :</b> " . Turkce::vasitaHali($noun) . "<br>";
        echo "<b>Sıra Sayı (Ordinal Number) :</b> " . Turkce::siraSayi($noun) . "<br>";
        echo "<b>Üleştirme (Distributive) :</b> " . Turkce::ulestirme($noun) . "<br>";
        echo "</blockquote>";
    }
}

?>
</body>
</html>
